### PR TITLE
Add EFO, DOID, and HP names and synonyms

### PR DESCRIPTION
**Pull request by cthoyt**
This piggybacks on the last PR to INDRA (https://github.com/sorgerlab/indra/pull/928) and uses the JSON resources generated by that new code to extract the names, identifiers, and synonyms for terms in three ontologies (via OBO): EFO, DOID, and HP.

### CODE REVIEW

File: gilda/generate_terms.py [hunk 1]
```diff
@@ -312,11 +347,18 @@ def get_all_terms():
     terms += generate_mesh_terms()
     terms += generate_uniprot_terms()
     terms += generate_adeft_terms()
+    terms += generate_doid_terms()
+    terms += generate_hp_terms()
+    terms += generate_efo_terms()
     return terms
 
 
-if __name__ == '__main__':
```
Comment (cthoyt) [inline gilda/generate_terms.py]: sticking all of this in a `main()` function reduces conflicts and makes it easier to turn this into an entrypoint later

Comment (bgyori) [reply gilda/generate_terms.py]: Yes, that's good

**Comment by cthoyt** [comment]
One thing we could do is iterate through all of the cross-references for the entry then add lots more terms, but with the alternate db and db_id (while keeping the source?) then there would need to be some post-processing logic on all of the terms to remove redundancies

**Comment by cthoyt** [comment]
Okay now I'm checking out what the consequences of adding this loop for the xrefs - it makes it slow down from < 1 second to mode than 30 minutes for each. Investigating now...

**Comment by cthoyt** [comment]
Lookup of MeSH identifiers solved with https://github.com/sorgerlab/indra/pull/933

**Comment by cthoyt** [comment]
Remaining missing DOID xrefs seem to be due to alternate identifiers. Looking into that now

**Comment by bgyori** [comment]
Alright, thanks again for this @cthoyt, I finally had a chance to work through this and merge it. I made the following changes:
- Migrated the code from the extra_terms file into the main generate_terms file.
- Reimplemented the xref mapping logic to apply once to a single entry (instead of generating out all xref combinations for each synonym) and replace it with a MESH or DOID ID. I reviewed the types of all the xrefs reported by all EFO, HP and DOID, and from what I can tell, only these two mappings are relevant in this context. I also got rid of mappings to supplementary MESH IDs which were often to outdated entries - in these cases we stick with the original source instead of MESH.
- Reviewed the entries we get from each source and found a number of corner cases that need to be post-processed or removed. There are still some weird synonyms but I handled most of the low-hanging-fruit issues.

**Comment by bgyori** [comment]
As a follow-up, not unexpectedly, there are redundancies not resolved by the source-provided mappings. For instance:
```
In [11]: ground('pancreatic cancer')                                                                                                          
Out[11]: 
[ScoredMatch(Term(pancreatic cancer,pancreatic cancer,DOID,DOID:1793,pancreatic cancer,name,efo),0.7777777777777778,Match(query=pancreatic cancer,ref=pancreatic cancer,exact=True,space_mismatch=False,dash_mismatches={},cap_combos=[])),
 ScoredMatch(Term(pancreatic cancer,pancreatic cancer,EFO,0002618,pancreatic carcinoma,synonym,efo),0.5555555555555556,Match(query=pancreatic cancer,ref=pancreatic cancer,exact=True,space_mismatch=False,dash_mismatches={},cap_combos=[])),
 ScoredMatch(Term(pancreatic cancer,pancreatic cancer,MESH,D010190,Pancreatic Neoplasms,synonym,efo),0.5555555555555556,Match(query=pancreatic cancer,ref=pancreatic cancer,exact=True,space_mismatch=False,dash_mismatches={},cap_combos=[]))]
```
In a separate PR, I will attempt to extend the automated MeSH mappings finder concept to these new redundancies.

File: gilda/generate_terms.py [hunk 2]
```diff
@@ -341,6 +352,113 @@ def generate_adeft_terms():
...
+
+def generate_efo_terms():
+    return _generate_obo_terms('efo')
+
+
+def generate_hp_terms():
+    return _generate_obo_terms('hp')
+
+
+def _generate_obo_terms(prefix):
```
Comment (cthoyt) [inline gilda/generate_terms.py:367]: I've written a *lot* more rules for normalizing terms and xrefs in the https://github.com/pyobo/pyobo repo. I'm still improving it but right now it loads all of the ontologies in OBO Foundry plus some ones I converted myself and extracts identifiers, labels, synonyms, and xrefs in a pretty general way. Throughout the process I've also started [curating a database](https://github.com/pyobo/pyobo/blob/master/src/pyobo/registries/metaregistry.json) of prefixes, synonyms for prefixes, etc. that would supplement the Identifiers.org database.

I don't think you'll want to have gilda depend on that code, but maybe you can get some ideas from it. Would be happy to explain further

**Pull request merged**